### PR TITLE
Attempt to shutdown web surfaces more consistently

### PR DESCRIPTION
**Pull request by jherico**
There have been a large number of web / QML related crashes in recent builds.  This PR is an attempt to resolve the issues by making the deconstruction of QML surfaces more consistent.

It makes a number of changes

* When destroying the surface for a web entity (displaying web content, not QML content) set the URL to an empty string before destruction.  This will hopefully trigger the backing chromium process to destroy any audio resources for the rendered page earlier.  See [FB 14104](https://highfidelity.manuscript.com/f/cases/14104/Audio-from-web-entity-continues-to-play-after-leaving-domain) and [FB 14380](https://highfidelity.manuscript.com/f/cases/14380/crash-in-shutdown-in-66-in-Browser-Audio-media-AudioManagerBase-AudioManagerBase)
* Eliminate `bool AbstractViewStateInterface::isAboutToQuit()`.  This function is probably not as reliable as we'd want it to be.  The problem is that many things in the application individually connect to the `QCoreApplication::aboutToQuit` signal, and there's no guarantee of the order in which those signal handlers will be called.  Because of this, code that's directly or indirectly running in response to `QCoreApplication::aboutToQuit` might call `bool AbstractViewStateInterface::isAboutToQuit()` and get a false negative because the `Application::onAboutToQuit` handler hasn't yet been called.  
* Add `void AbstractViewStateInterface::sendLambdaEvent(const std::function<void()>& f)`.  This function is similar to the `AbstractViewStateInterface::postLambdaEvent` except that it's guaranteed not to return until the code in the lambda has executed, regardless of what thread it's called from.  
* Modify the code in Web3DOverlay and WebEntityRenderer to delete their QML surfaces in a more consistent fashion.  The removal of `isAboutToQuit` means these classes will now always do an immediate delete of the QML surface, rather than deferring it with a call to `deleteLater`.  Further, the code uses the new `AbstractViewStateInterface::sendLambdaEvent` to ensure that regardless of what thread pointer is destroyed on, the thread will stop and wait for the QML surface to be deleted before continuing.  In theory this should not affect the application, since the intent of the current design is that the deleter is only ever called on the main thread.  However, if there is an edge case where the deleter is being called on a background thread, this change should help us produce better stack traces to identify why this is happening.  
* Refactor `hifi::qml::OffscreenSurface` shutdown logic, using the Qt multi-threaded QQuickControl [example](http://doc.qt.io/qt-5/qtquick-rendercontrol-window-multithreaded-cpp.html) as a template

## Testing

Try to reproduce any of ...
* [FB 14104](https://highfidelity.manuscript.com/f/cases/14104/Audio-from-web-entity-continues-to-play-after-leaving-domain)
* [FB 14380](https://highfidelity.manuscript.com/f/cases/14380/crash-in-shutdown-in-66-in-Browser-Audio-media-AudioManagerBase-AudioManagerBase)
* [FB 14381](https://highfidelity.manuscript.com/f/cases/14381/crash-on-shutdown-in-QQuickRenderControl-QQuickRenderControl)
* [FB 14382](https://highfidelity.manuscript.com/f/cases/14382/crash-on-exit-in-hifi-qml-OffscreenSurface-OffscreenSurface) 

...with this build.  

In particular, try going to domains that contain WebEntity content, especially those with audio or a/v playback, and test entering and leaving the domain repeatedly, as well as entering the domain and then exiting the application while a web entity is in view and playing audio or video.  

Testing should focus on lower end machines as they are the most likely to encounter issues arising from timing problems or race conditions with multiple threads.  

Please also test on Mac, as there are reports that this platform always crashes on exit.

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/9040/app_12969_PR_fd014a5f74a46d80effd7a2938715216494cd74c.apk

**Comment by divyaelango88** [comment]
I did a quick 5 minute sanity testing on Android and it does not break it. So removing the android testing required tag.

File: libraries/qml/src/qml/impl/SharedObject.cpp [hunk 1]
```diff
@@ -72,26 +72,35 @@ SharedObject::SharedObject() {
...
+        _quickWindow->destroy();
+        delete _quickWindow;
+        _quickWindow = nullptr;
     }
 
-    if (_rootItem) {
-        _rootItem->deleteLater();
-        _rootItem = nullptr;
-    }
+    // _rootItem is parented to the quickWindow, so needs no explicit destruction
```
Comment (samcake) [inline libraries/qml/src/qml/impl/SharedObject.cpp]: comments?

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/9087/app_12969_PR_ecaf162afb1c0e7a5abda6c4bde1e034ea28e061.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/9087/HighFidelity-Beta-PR12969-ecaf162afb1c0e7a5abda6c4bde1e034ea28e061.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/9087/HighFidelity-Beta-Interface-PR12969-ecaf162afb1c0e7a5abda6c4bde1e034ea28e061.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/9087/HighFidelity-Beta-PR12969-ecaf162afb1c0e7a5abda6c4bde1e034ea28e061.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/9087/HighFidelity-Beta-Interface-PR12969-ecaf162afb1c0e7a5abda6c4bde1e034ea28e061.exe

**Comment by qlimnavarro** [comment]
Currently Testing

**Pull request merged**